### PR TITLE
[Serve] Add status code to retry when request timed out

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

It's not immediately apparent what's the status code returned by Serve when the request timed out. Added a note section on the timeout setup so user knows what status code to expect and potentially retry.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by edoakes — APPROVED** [review]

**Review by akshay-anyscale — APPROVED** [review]

File: doc/source/serve/advanced-guides/performance.md [hunk 1]
```diff
@@ -55,7 +55,16 @@ proper backpressure. You can increase the value in the deployment decorator; e.g
...
+You can set an end-to-end timeout for HTTP requests by setting the `request_timeout_s`
+in the `http_options` field of the Serve config. HTTP Proxies wait for that many
+seconds before terminating an HTTP request. This config is global to your Ray cluster,
+and it can't be updated during runtime. Use [client-side retries](serve-best-practices-http-requests)
+to retry requests that time out due to transient failures.
+
+:::{note}
+Serve returns response with status code `408` when the request timed out. Clients can
+handle this status code and retry the request.
+:::
```
Comment (shrekris-anyscale) [inline doc/source/serve/advanced-guides/performance.md:67]:
```suggestion
:::{note}
Serve returns a response with status code `408` when a request times out. Clients can retry when they receive this `408` response.
:::
```

**Review by shrekris-anyscale — COMMENTED** [review]
Should clients retry `408` responses? 4xx responses usually mean that something about the request was wrong.

For example, maybe the client submitted a very large image that takes too long to process. Retrying that wouldn't resolve the `408`, and it would consume more compute resources.

File: doc/source/serve/advanced-guides/performance.md (hunk 1 above)
Comment (GeneDer) [reply doc/source/serve/advanced-guides/performance.md]: Yea, I don't think the client have to retry every 408. And good point about retrying long running processing likely to timed out again. But the goal is really to let the user know what status code to expect and handle accordingly.

**Comment by shrekris-anyscale** [comment]
It looks like other services do recommend retrying on `408` (e.g. [Google Cloud Storage](https://cloud.google.com/storage/docs/retry-strategy#retryable)). I'm okay with the recommendation as-is.

If we see more users run into `408` errors in the future due to malformed requests, we can add a caveat to the note.

**Review by shrekris-anyscale — APPROVED** [review]

File: doc/source/serve/advanced-guides/performance.md [hunk 2]
```diff
@@ -55,7 +55,15 @@ proper backpressure. You can increase the value in the deployment decorator; e.g
 
 By default, Serve lets client HTTP requests run to completion no matter how long they take. However, slow requests could bottleneck the replica processing, blocking other requests that are waiting. Set an end-to-end timeout, so slow requests can be terminated and retried.
 
-You can set an end-to-end timeout for HTTP requests by setting the `request_timeout_s` in the `http_options` field of the Serve config. HTTP Proxies wait for that many seconds before terminating an HTTP request. This config is global to your Ray cluster, and it can't be updated during runtime. Use [client-side retries](serve-best-practices-http-requests) to retry requests that time out due to transient failures.
+You can set an end-to-end timeout for HTTP requests by setting the `request_timeout_s`
```
Comment (angelinalg) [inline doc/source/serve/advanced-guides/performance.md:58]:
```suggestion
You can set an end-to-end timeout for HTTP requests by setting the `request_timeout_s` parameter
```

File: doc/source/serve/advanced-guides/performance.md [hunk 3]
```diff
@@ -55,7 +55,15 @@ proper backpressure. You can increase the value in the deployment decorator; e.g
...
+in the `http_options` field of the Serve config. HTTP Proxies wait for that many
+seconds before terminating an HTTP request. This config is global to your Ray cluster,
+and it can't be updated during runtime. Use [client-side retries](serve-best-practices-http-requests)
```
Comment (angelinalg) [inline doc/source/serve/advanced-guides/performance.md:61]:
```suggestion
and you can't update it during runtime. Use [client-side retries](serve-best-practices-http-requests)
```

**Review by angelinalg — APPROVED** [review]
Just a couple of style nits.

**Comment by GeneDer** [comment]
@edoakes this is ready for merge

**Pull request merged**